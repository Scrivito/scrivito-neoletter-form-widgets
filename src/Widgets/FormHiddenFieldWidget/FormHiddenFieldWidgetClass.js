import * as Scrivito from "scrivito";

export const FormHiddenFieldWidget = Scrivito.provideWidgetClass(
  "FormHiddenFieldWidget",
  {
    onlyInside: ["FormContainerWidget"],
    attributes: {
      customFieldName: "string",
      hiddenValue: "string",
    },
  }
);
