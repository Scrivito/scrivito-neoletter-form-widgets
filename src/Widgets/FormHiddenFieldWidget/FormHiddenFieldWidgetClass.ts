import * as Scrivito from "scrivito";

export const FormHiddenFieldWidget = Scrivito.provideWidgetClass(
  "FormHiddenFieldWidget",
  {
    onlyInside: [
      "FormStepContainerWidget",
      "FormStepWidget",
      "FormContainerWidget"
    ],
    attributes: {
      customFieldName: "string",
      hiddenValue: "string",
      type: [
        "enum",
        {
          values: ["custom", "subscription"]
        }
      ]
    }
  }
);
