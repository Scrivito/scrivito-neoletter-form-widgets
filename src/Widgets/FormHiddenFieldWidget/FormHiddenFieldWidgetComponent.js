import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../FormContainerWidget/utils/getFieldName";

Scrivito.provideComponent("FormHiddenFieldWidget", ({ widget }) => {
  const name = getFieldName(widget);
  if (!name) return null;

  return <input type="hidden" name={name} value={widget.get("hiddenValue")} />;
});
