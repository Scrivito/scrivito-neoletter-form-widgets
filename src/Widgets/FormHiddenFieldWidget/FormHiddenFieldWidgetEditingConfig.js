import * as Scrivito from "scrivito";
import formHiddenFieldWidgetIcon from "../../assets/images/form_hidden_field_widget.svg";
import { customFieldNameValidation } from "../FormContainerWidget/utils/validations/customFieldNameValidation";
//TODO: add/check insideFormValidation
Scrivito.provideEditingConfig("FormHiddenFieldWidget", {
  title: "Hidden Form Field",
  thumbnail: formHiddenFieldWidgetIcon,
  attributes: {
    customFieldName: { title: "Field name" },
    hiddenValue: {
      title: "Hidden value",
      description: "This value is sent on every submission of the form.",
    },
  },
  properties: ["customFieldName", "hiddenValue"],
  initialContent: {
    customFieldName: "custom_hidden_field",
  },
  validations: [
    customFieldNameValidation,
    (widget) => {
      if (
        widget
          .container()
          .get("hiddenFields")
          .map((w) => w.id())
          .includes(widget.id())
      ) {
        return;
      }

      return {
        message: "Hidden fields should be added in the properties of the form.",
        severity: "info",
      };
    },
  ],
  titleForContent: (widget) =>
    `Hidden Form Field: ${[
      widget.get("customFieldName"),
      widget.get("hiddenValue"),
    ]
      .filter((e) => e)
      .join(" - ")}`,
});
