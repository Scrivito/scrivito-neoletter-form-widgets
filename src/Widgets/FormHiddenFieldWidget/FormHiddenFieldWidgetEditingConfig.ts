import * as Scrivito from "scrivito";
import formHiddenFieldWidgetIcon from "../../assets/images/form_hidden_field_widget.svg";
import { customFieldNameValidation } from "../FormStepContainerWidget/utils/validations/customFieldNameValidation";
import { getFormContainer } from "../FormStepContainerWidget/utils/getFormContainer";
import { isCustomType } from "../FormStepContainerWidget/utils/isCustomType";
import { getFieldName } from "../FormStepContainerWidget/utils/getFieldName";
import { typeValidation } from "../FormStepContainerWidget/utils/validations/typeValidation";

Scrivito.provideEditingConfig("FormHiddenFieldWidget", {
  title: "Neoletter Hidden Form Field",
  thumbnail: formHiddenFieldWidgetIcon,
  attributes: {
    customFieldName: { title: "Field name" },
    hiddenValue: {
      title: "Hidden value",
      description: "This value is sent on every submission of the form."
    },
    type: {
      title: "Input type",
      values: [
        { value: "custom", title: "Custom" },
        { value: "subscription", title: "Subscription" }
      ]
    }
  },
  properties: (widget: Scrivito.Widget) => {
    if (!isCustomType(widget)) {
      return ["type", "hiddenValue"];
    }
    return ["type", "customFieldName", "hiddenValue"];
  },
  initialContent: {
    customFieldName: "custom_hidden_field",
    type: "custom"
  },
  validations: [
    customFieldNameValidation,
    typeValidation,
    (widget: Scrivito.Widget) => {
      const container = getFormContainer(widget);
      if (container && container.get("hiddenFields")) {
        const hiddenFields = container.get("hiddenFields") as Scrivito.Widget[];
        if (
          hiddenFields &&
          hiddenFields.map(w => w.id()).includes(widget.id())
        ) {
          return;
        }
      }
      return {
        message: "Hidden fields should be added in the properties of the form.",
        severity: "info"
      };
    },
    [
      "hiddenValue",
      (hiddenValue: string, { widget }: { widget: Scrivito.Widget }) => {
        const fieldName = getFieldName(widget);

        if (fieldName === "subscription" && hiddenValue !== "on") {
          return {
            message:
              "Please enter 'on' to activate the subscription process on every submission.",
            severity: "warning"
          };
        }
      }
    ]
  ],

  titleForContent: widget => {
    const fieldName = getFieldName(widget);
    return `Hidden Form Field: ${[fieldName, widget.get("hiddenValue")]
      .filter(e => e)
      .join(" - ")}`;
  }
});
