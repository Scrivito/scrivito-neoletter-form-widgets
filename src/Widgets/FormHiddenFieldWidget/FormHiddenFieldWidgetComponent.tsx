import * as React from "react";
import * as Scrivito from "scrivito";
import { isEmpty } from "lodash-es";
import { getFieldName } from "../FormStepContainerWidget/utils/getFieldName";
import { FormHiddenFieldWidget } from "./FormHiddenFieldWidgetClass";

Scrivito.provideComponent(FormHiddenFieldWidget, ({ widget }) => {
  const name = getFieldName(widget);
  if (isEmpty(name)) return null;
  return <input type="hidden" name={name} value={widget.get("hiddenValue")} />;
});
