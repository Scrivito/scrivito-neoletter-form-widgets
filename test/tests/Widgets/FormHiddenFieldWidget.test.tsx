import * as Scrivito from "scrivito";
import { FormHiddenFieldWidget } from "../../../src/Widgets/FormHiddenFieldWidget/FormHiddenFieldWidgetClass";
import PageRenderer from "../../helpers/pageRenderer";
import "../../../src/Widgets/FormHiddenFieldWidget/FormHiddenFieldWidgetComponent";

Scrivito.configure({ tenant: "inMemory" });

const pageRenderer = new PageRenderer();

const widgetProps = {
  customFieldName: "custom_hidden",
  hiddenValue: "hidden",
  type: "custom"
};
describe("FormHiddenFieldWidget", () => {
  it("renders null when name is empty", () => {
    pageRenderer.render({
      body: [new FormHiddenFieldWidget({ ...widgetProps, customFieldName: "" })]
    });
    const hiddenInput = document.querySelector('[type="hidden"]')!;
    expect(hiddenInput).not.toBeInTheDocument();
  });

  it("renders input field when name is not empty", () => {
    pageRenderer.render({
      body: [new FormHiddenFieldWidget(widgetProps)]
    });
    const hiddenInput = document.querySelector('[type="hidden"]')!;
    expect(hiddenInput).toBeInTheDocument();
    expect(hiddenInput).toHaveAttribute("value", "hidden");
  });

  it("renders correctly", () => {
    const tree = pageRenderer.getAsJSON({
      body: [new FormHiddenFieldWidget(widgetProps)]
    });
    expect(tree).toMatchSnapshot();
  });
});
