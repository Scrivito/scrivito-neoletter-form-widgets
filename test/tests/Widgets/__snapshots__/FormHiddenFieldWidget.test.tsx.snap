// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormHiddenFieldWidget renders correctly 1`] = `
<div>
  <div>
    <input
      name={null}
      type="hidden"
      value="hidden"
    />
  </div>
</div>
`;
